### PR TITLE
Reduce visibility of `g_openmp_hardware_max_threads` symbol to avoid trouble with shared libabries

### LINKED ISSUE #7283: Building with pybind11 can't properly intilize Kokkos::Impl::g_openmp_hardware_max_threads
**Describe the bug**

When building Python module using pybind11 and linking to Kokkos with OpenMP backend, the variable `Kokkos::Impl::g_openmp_hardware_max_threads` is always `1`, which breaks `ScatterView`.

**Please include the following for a minimal reproducer**

- Platform: Ubuntu 22.04 on Intel 9700k
- Compiler: clang++ 17.0.6
- Kokkos version: 4.4.0 and commit a068d826dd79e8994c7b43a3b72577e0ea704b8f
- pybind11 version: 2.12.0dev1
- CMake configure command: `CXX=clang++ cmake .. -DKokkos_ENABLE_OPENMP=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON` (`BUILD_SHARED_LIBS=OFF` results in linkage error)

C++ file (`pytest.cpp`)
```cpp
#include <Kokkos_Core.hpp>
#include <pybind11/pybind11.h>
#include <omp.h>

void initialize() {
    printf("Value of g_openmp_hardware_max_threads before initialize is %d\n", Kokkos::Impl::g_openmp_hardware_max_threads);
    Kokkos::initialize();
    printf("Value of g_openmp_hardware_max_threads after initialize is %d, while omp_current_max_threads is %d\n",
        Kokkos::Impl::g_openmp_hardware_max_threads,
        omp_get_max_threads()
    );
}

void finalize() {
    printf("Finalizing\n");
    Kokkos::finalize();
}

PYBIND11_MODULE(pytest, m) {
    m.def("initialize", &initialize);
    m.def("finalize", &finalize);
}
```
CMakeLists.txt:
```cmake
cmake_minimum_required (VERSION 3.10)
project(pytest)

add_subdirectory(kokkos)
find_package(pybind11 REQUIRED)

pybind11_add_module(pytest pytest.cpp)
target_link_libraries(pytest PUBLIC Kokkos::kokkos)
```

test.py
```python
import pytest

pytest.initialize()
pytest.finalize()
```

Output of `python3 test.py`:
```
Value of g_openmp_hardware_max_threads before initialize is 1
Value of g_openmp_hardware_max_threads after initialize is 1, while omp_current_max_threads is 8
Finalizing
```

If add a print line in this function: https://github.com/kokkos/kokkos/blob/08ceff92bcf3a828844480bc1e6137eb74028517/core/src/OpenMP/Kokkos_OpenMP_Instance.cpp#L161
the printed value of `Kokkos::Impl::g_openmp_hardware_max_threads` is correct (equal to `omp_get_max_threads`).

Also, changing the declaration of this variable from `inline` to `extern` in the header file and defining it in `core/src/OpenMP/Kokkos_OpenMP_Instance.cpp` fix the issue.

### PR DESCRIPTION
**Pull request by dalg24**
Fix #7283 (not tested)

### CODE REVIEW

File: core/src/OpenMP/Kokkos_OpenMP_UniqueToken.hpp [hunk 1]
```diff
@@ -105,7 +105,8 @@ class UniqueToken<OpenMP, UniqueTokenScope::Global> {
   /// \brief upper bound for acquired values, i.e. 0 <= value < size()
   KOKKOS_INLINE_FUNCTION
   int size() const noexcept {
-    KOKKOS_IF_ON_HOST((return Kokkos::Impl::g_openmp_hardware_max_threads;))
+    KOKKOS_IF_ON_HOST(
+        (return Kokkos::Impl::OpenMPInternal::max_hardware_threads();))
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_UniqueToken.hpp:109]: Note that this now would involve a function call.
It looks like this is fine as far as `ScatterView` is concerned (not used in the hot path as far as I can tell) but we probably should discuss it a bit.

**Review by crtrott — APPROVED** [review]
Fine with me assuming it fixes the issue. My gut feeling is that this should not be a hot path issue with UniqueToken.

File: core/src/OpenMP/Kokkos_OpenMP_Instance.cpp [hunk 2]
```diff
@@ -188,9 +196,9 @@ void OpenMPInternal::initialize(int thread_count) {
     // Before any other call to OMP query the maximum number of threads
     // and save the value for re-initialization unit testing.
 
-    Impl::g_openmp_hardware_max_threads = get_current_max_threads();
+    g_openmp_hardware_max_threads = get_current_max_threads();
 
-    int process_num_threads = Impl::g_openmp_hardware_max_threads;
+    int process_num_threads = g_openmp_hardware_max_threads;
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Instance.cpp:201]: Wouldn't it make sense to only set this for the default execution space instance?

Comment (dalg24) [reply core/src/OpenMP/Kokkos_OpenMP_Instance.cpp:201]: I think so.  But such a refactor/fix is beyond the scope of this PR.

**Review by masterleinad — APPROVED** [review]

**Comment by masterleinad** [comment]
I verified that we see the test failure with OpenMP when building with shared libraries (and `-fvisibility=hidden`) in user code.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please

**Comment by dalg24** [comment]
Unrelated failure in one of the CUDA build (GPU not detected at runtime)

**Pull request merged**